zcash_client_backend/src/scan.rs

```diff
@@ -1,200 +1,19 @@
-    fn activation_height(&self, nu: consensus::NetworkUpgrade) -> Option<consensus::BlockHeight> {
-        match self {
-            Self::Consensus(params) => params.activation_height(nu),
-            Self::RegTest {
-                overwinter,
-                sapling,
-                blossom,
-                heartwood,
-                canopy,
-                nu5,
-            } => match nu {
-                consensus::NetworkUpgrade::Overwinter => *overwinter,
-                consensus::NetworkUpgrade::Sapling => *sapling,
-                consensus::NetworkUpgrade::Blossom => *blossom,
-                consensus::NetworkUpgrade::Heartwood => *heartwood,
-                consensus::NetworkUpgrade::Canopy => *canopy,
-                consensus::NetworkUpgrade::Nu5 => *nu5,
-            },
-        }
-    }
-
-    fn coin_type(&self) -> u32 {
-        match self {
-            Self::Consensus(params) => params.coin_type(),
-            Self::RegTest { .. } => constants::regtest::COIN_TYPE,
-        }
-    }
-
-    fn hrp_sapling_extended_spending_key(&self) -> &str {
-        match self {
-            Self::Consensus(params) => params.hrp_sapling_extended_spending_key(),
-            Self::RegTest { .. } => constants::regtest::HRP_SAPLING_EXTENDED_SPENDING_KEY,
-        }
-    }
-
-    fn hrp_sapling_extended_full_viewing_key(&self) -> &str {
-        match self {
-            Self::Consensus(params) => params.hrp_sapling_extended_full_viewing_key(),
-            Self::RegTest { .. } => constants::regtest::HRP_SAPLING_EXTENDED_FULL_VIEWING_KEY,
-        }
-    }
-
-    fn hrp_sapling_payment_address(&self) -> &str {
-        match self {
-            Self::Consensus(params) => params.hrp_sapling_payment_address(),
-            Self::RegTest { .. } => constants::regtest::HRP_SAPLING_PAYMENT_ADDRESS,
-        }
-    }
-
-    fn b58_pubkey_address_prefix(&self) -> [u8; 2] {
-        match self {
-            Self::Consensus(params) => params.b58_pubkey_address_prefix(),
-            Self::RegTest { .. } => constants::regtest::B58_PUBKEY_ADDRESS_PREFIX,
-        }
-    }
-
-    fn b58_script_address_prefix(&self) -> [u8; 2] {
-        match self {
-            Self::Consensus(params) => params.b58_script_address_prefix(),
-            Self::RegTest { .. } => constants::regtest::B58_SCRIPT_ADDRESS_PREFIX,
-        }
-    }
-}
+use zcash_note_encryption::{batch, BatchDomain, Domain, ShieldedOutput, COMPACT_NOTE_SIZE};
+use zcash_primitives::{block::BlockHash, transaction::TxId};
 
 /// A decrypted note.
-struct DecryptedNote<D: Domain> {
+pub(crate) struct DecryptedNote<D: Domain> {
     /// The incoming viewing key used to decrypt the note.
-    ivk: D::IncomingViewingKey,
+    pub(crate) ivk: D::IncomingViewingKey,
     /// The recipient of the note.
-    recipient: D::Recipient,
+    pub(crate) recipient: D::Recipient,
     /// The note!
-    note: D::Note,
-    /// The memo sent with the note.
-    memo: D::Memo,
```

Comment:
Non-blocking: we should figure out how to generalise this logic over compact and non-compact outputs, so we can potentially share it with `zcashd` (assuming the logic doesn't change significantly when we move to async).

Reply:
Agreed; I made a quick attempt at it before we started pairing on this but it added enough complexity that think that should be a separate refactoring.